File: apps/mobile/lib/domain/entities/product.dart
class Product {
  final String id;
  final String title;
  final String? brand;
  final double? price;
  final String? currency;
  final double? rating;
  final bool isDemoData;

  const Product({
    required this.id,
    required this.title,
    required this.brand,
    required this.price,
    required this.currency,
    required this.rating,
    required this.isDemoData,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String,
      title: json['title'] as String,
      brand: json['brand'] as String?,
      price: (json['price'] as num?)?.toDouble(),
      currency: json['currency'] as String?,
      rating: (json['rating'] as num?)?.toDouble(),
      isDemoData: json['isDemoData'] as bool? ?? false,
    );
  }
}

class Deal {
  final String id;
  final String merchant;
  final double price;
  final double? originalPrice;
  final double? discountPercent;
  final String currency;

  const Deal({
    required this.id,
    required this.merchant,
    required this.price,
    required this.originalPrice,
    required this.discountPercent,
    required this.currency,
  });

  factory Deal.fromJson(Map<String, dynamic> json) {
    return Deal(
      id: json['id'] as String,
      merchant: json['merchant'] as String,
      price: (json['price'] as num).toDouble(),
      originalPrice: (json['originalPrice'] as num?)?.toDouble(),
      discountPercent: (json['discountPercent'] as num?)?.toDouble(),
      currency: json['currency'] as String,
    );
  }
}

File: packages/contracts/src/product.ts
import { z } from "zod";

export const Product = z.object({
  id: z.string(),
  title: z.string(),
  brand: z.string().nullable(),
  category: z.string(),
  price: z.number().nonnegative().nullable(),
  currency: z.string().length(3).nullable(),
  rating: z.number().min(0).max(5).nullable(),
  reviewCount: z.number().int().nonnegative().nullable(),
  availability: z.enum(["IN_STOCK", "OUT_OF_STOCK", "LIMITED", "UNKNOWN"]),
  seller: z.string().nullable(),
  source: z.string(),
  sourceProductId: z.string(),
  warranty: z.string().nullable(),
  returnInformation: z.string().nullable(),
  isDemoData: z.boolean().default(false),
  fetchedAt: z.string().datetime(),
});
export type Product = z.infer<typeof Product>;

export const ProductAnalysis = z.object({
  productId: z.string(),
  pros: z.array(z.string()),
  cons: z.array(z.string()),
  reviewThemes: z.array(z.string()),
  valueConsiderations: z.array(z.string()),
  alternativeProductIds: z.array(z.string()),
  generatedAt: z.string().datetime(),
});
export type ProductAnalysis = z.infer<typeof ProductAnalysis>;

export const Deal = z.object({
  id: z.string(),
  productId: z.string(),
  merchant: z.string(),
  price: z.number().nonnegative(),
  originalPrice: z.number().nonnegative().nullable(),
  discountPercent: z.number().min(0).max(100).nullable(),
  currency: z.string().length(3),
  validFrom: z.string().datetime(),
  validUntil: z.string().datetime(),
  source: z.string(),
  verifiedAt: z.string().datetime(),
});
export type Deal = z.infer<typeof Deal>;

export function isDealActive(deal: Pick<Deal, "validFrom" | "validUntil">, now: Date = new Date()): boolean {
  const nowMs = now.getTime();
  return new Date(deal.validFrom).getTime() <= nowMs && nowMs <= new Date(deal.validUntil).getTime();
}
